benchmarks/exp_hash.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2025 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Script for running a benchmark to pick a hashing algorithm."""
+
+import argparse
+import timeit
+from typing import Final
+
+import numpy as np
+
+from model_signing.hashing import hashing
+from model_signing.hashing import memory
+
+
+KB: Final[int] = 1024
+MB: Final[int] = 1024 * KB
+GB: Final[int] = 1024 * MB
+
+
+def build_parser() -> argparse.ArgumentParser:
+    """Builds the command line parser for the hash experiment."""
+    parser = argparse.ArgumentParser(
+        description="hash algorithm benchmark data for model signing"
+    )
+
+    parser.add_argument(
+        "--repeat",
+        help="how many times to repeat each algorithm",
+        type=int,
+        default=5,
+    )
+
+    parser.add_argument(
+        "--methods",
+        help="hash methods to benchmark",
+        nargs="+",
+        type=str,
+        default=["sha256", "blake2"],
+    )
+
+    parser.add_argument(
+        "--data-sizes",
+        help="hash methods to benchmark",
+        nargs="+",
+        type=int,
+        default=[KB, MB, 512 * MB, GB, 4 * GB, 16 * GB, 32 * GB],
+    )
+
+    return parser
+
+
+def _human_size(size: int) -> str:
+    if size >= GB:
+        return str(size / GB) + " GB"
+    elif size >= MB:
+        return str(size / MB) + " MB"
+    elif size >= KB:
+        return str(size / KB) + " KB"
+    return str(size) + " B"
+
+
+def _get_hasher(hash_algorithm: str) -> hashing.StreamingHashEngine:
+    match hash_algorithm:
+        case "sha256":
+            return memory.SHA256()
+        case "blake2":
+            return memory.BLAKE2()
+    raise ValueError(f"Cannot convert {hash_algorithm} to a hash engine")
+
+
+def _generate_data(size: int) -> bytes:
+    if size < 0:
+        raise ValueError("Cannot generate negative bytes")
+    return np.random.randint(0, 256, size, dtype=np.uint8).tobytes()
+
+
+if __name__ == "__main__":
+    np.random.seed(42)
+    args = build_parser().parse_args()
+    data = _generate_data(max(args.data_sizes))
+    for size in args.data_sizes:
+        for algorithm in args.methods:
+            hasher = _get_hasher(algorithm)
+
+            def hash(hasher=hasher, size=size):
+                hasher.update(data[:size])
```

Comment:
Should we reinitialize the hasher too under the measured scope? We can make `_get_hasher` return just the constructor and call it here

Reply:
since this is in the inner most loop, it's always a new hasher. I dont think we need to reset anything

Reply:
But each call of `hash` from `timeit` just hashes data, it doesn't time the time it takes to init the hasher.